### ISSUE TITLE
opte kernel panic

### ISSUE THREAD

**Issue by jmpesp** [oxidecomputer/opte#335]
Saw the following panic:

```
james@kibblesnbits:~$ fmdump -Vp -u ecd5a6f0-6c1d-e74d-eeee-d5a29265d851
TIME                           UUID                                 SUNW-MSG-ID
Feb 17 2023 20:59:12.711862000 ecd5a6f0-6c1d-e74d-eeee-d5a29265d851 SUNOS-8000-KL

  TIME                 CLASS                                 ENA
  Feb 17 20:59:12.6936 ireport.os.sunos.panic.dump_available 0x0000000000000000
  Feb 17 20:58:57.6792 ireport.os.sunos.panic.dump_pending_on_device 0x0000000000000000

nvlist version: 0
        version = 0x0
        class = list.suspect
        uuid = ecd5a6f0-6c1d-e74d-eeee-d5a29265d851
        code = SUNOS-8000-KL
        diag-time = 1676667552 695222
        de = fmd:///module/software-diagnosis
        fault-list-sz = 0x1
        fault-list = (array of embedded nvlists)
        (start fault-list[0])
        nvlist version: 0
                version = 0x0
                class = defect.sunos.kernel.panic
                certainty = 0x64
                asru = sw:///:path=/var/crash/kibblesnbits/.ecd5a6f0-6c1d-e74d-eeee-d5a29265d851
                resource = sw:///:path=/var/crash/kibblesnbits/.ecd5a6f0-6c1d-e74d-eeee-d5a29265d851
                savecore-succcess = 1
                dump-dir = /var/crash/kibblesnbits
                dump-files = vmdump.1
                os-instance-uuid = ecd5a6f0-6c1d-e74d-eeee-d5a29265d851
                panicstr = panicked at 'called `Result::unwrap()` on an `Err` value: PacketSeg { mp: 0xfffffeb1ddc0ec20, dblk: 0xfffffeb1ddd38d00, len: 68, avail: 208 }', /work/oxidecomputer/opte/opte/src/engine/packet.rs:801:
                panicstack = xde:rust_begin_unwind+146 () | xde:_ZN4core9panicking9panic_fmt17hbd42276bcaa2dfd8E+2d () | xde:_ZN4core6result13unwrap_failed17h5b67be868f6dc91bE+6c () | xde:_ZN4opte6engine6packet47Packet$LT$opte..engine..packet..Initialized$GT$9wrap_mblk17h39b3c73afaee430dE+13b () | xde:xde_rx+47 () | mac:mac_promisc_dispatch_one+ac () | mac:mac_promisc_dispatch+82 () | mac:mac_provider_tx+57 () | mac:mac_tx_send+288 () | mac:mac_tx_soft_ring_process+89 () | mac:mac_tx_fanout_mode+11e () | mac:mac_tx+1a9 () | dld:str_mdata_fastpath_put+8e () | dld:dld_wput+1ac () | unix:putnext+233 () | ip:ip_xmit+632 () | ip:ip_forward_xmit_v6+f2 () | ip:ire_recv_forward_v6+3c2 () | ip:ill_input_short_v6+472 () | ip:ip_input_common_v6+283 () | ip:ip_input_v6+1f () | ip:ip_rput_v6+71 () | unix:putnext+233 () | dld:dld_str_rx_fastpath+37 () | dls:i_dls_link_rx+311 () | mac:mac_rx_deliver+2e () | mac:mac_rx_soft_ring_process+1be () | mac:mac_rx_srs_proto_fanout+29a () | mac:mac_rx_srs_drain+2e5 () | mac:mac_rx_srs_process+123 () | mac:mac_tx_send+269 () | mac:mac_tx_single_ring_mode+82 () | mac:mac_tx+1a9 () | dld:str_mdata_fastpath_put+8e () | dld:dld_wput+1ac () | unix:putnext+233 () | ip:ip_xmit+632 () | ip:ire_send_wire_v6+126 () | ip:conn_ip_output+1d4 () | ip:tcp_send_data+58 () | ip:tcp_send+8d2 () | ip:tcp_wput_data+68a () | ip:tcp_input_data+730 () | ip:squeue_enter+409 () | ip:ip_fanout_v6+9bb () | ip:ip_input_local_v6+1e () | ip:ire_recv_local_v6+131 () | ip:ill_input_short_v6+472 () | ip:ip_input_common_v6+283 () | ip:ip_input_v6+1f () | ip:ip_rput_v6+71 () | unix:putnext+233 () | dld:dld_str_rx_fastpath+37 () | dls:i_dls_link_rx+311 () | mac:mac_rx_deliver+2e () | mac:mac_rx_soft_ring_process+1be () | mac:mac_rx_srs_proto_fanout+29a () | mac:mac_rx_srs_drain+2e5 () | mac:mac_rx_srs_process+123 () | mac:mac_tx_send+269 () | mac:mac_tx_single_ring_mode+82 () | mac:mac_tx+1a9 () | dld:str_mdata_fastpath_put+8e () | dld:dld_wput+1ac () | unix:putnext+233 () | ip:ip_xmit+632 () | ip:ip_forward_xmit_v6+f2 () | 
                crashtime = 1676667422
                panic-time = Fri Feb 17 20:57:02 2023 UTC
        (end fault-list[0])

        fault-status = 0x1
        severity = Major
        __ttl = 0x1
        __tod = 0x63efeaa0 0x2a6e26f0
```

Here's some system details:

```
james@kibblesnbits:~$ uname -a
SunOS kibblesnbits 5.11 helios-1.0.21604 i86pc i386 i86pc

james@kibblesnbits:~$ pkg info pkg://helios-netdev/driver/network/opte
             Name: driver/network/opte
          Summary: The Oxide Packet Transformation Engine
         Category: Drivers/Networking
            State: Installed (Manually installed)
        Publisher: helios-netdev
          Version: 0.19.152
           Branch: None
   Packaging Date: January 10, 2023 at 12:29:33 AM
Last Install Time: August 19, 2022 at 04:35:04 PM
 Last Update Time: February 17, 2023 at 07:15:27 PM
             Size: 38.64 MB
             FMRI: pkg://helios-netdev/driver/network/opte@0.19.152:20230110T002933Z
             
james@kibblesnbits:~$ pkg info pkg://on-nightly/consolidation/osnet/osnet-incorporation
             Name: consolidation/osnet/osnet-incorporation
          Summary: OS/Net consolidation incorporation
      Description: This incorporation constrains packages from the OS/Net
                   consolidation.
            State: Installed (Manually installed)
        Publisher: on-nightly
          Version: 0.5.11
           Branch: 1.0.21604
   Packaging Date: December  6, 2022 at 07:09:25 PM
Last Install Time: September  8, 2021 at 06:52:19 AM
 Last Update Time: February 17, 2023 at 07:15:51 PM
             Size: 0.00 B
             FMRI: pkg://on-nightly/consolidation/osnet/osnet-incorporation@0.5.11-1.0.21604:20221206T190925Z
```

**Comment by jmpesp** [comment]
```
james@kibblesnbits:~$ modinfo | grep xde
218 fffffffff8196000 11b730 118   1  xde (xde)
```

**Comment by bnaecker** [comment]
Looks like you've received an mblk chain, and that chain contains more mblks than the [current limit of 4](https://github.com/oxidecomputer/opte/blob/1b16a452243e9d7d686e8c84b4ff4f3305acc061/opte/src/engine/packet.rs#L369). I'm not familiar with these update Ryan did, but we'll need to bump that static size (to what, I'm not sure), and obviously remove the unwrap.

**Comment by bnaecker** [comment]
@jmpesp Can you place the dump somewhere? It'll be helpful to dig around and see what the actual size of the chain was for example.

**Comment by jmpesp** [comment]
Yeah, I'm uploading it somewhere now

**Comment by bnaecker** [comment]
I was going to ask about jumbo frames, so that's helpful. I'm a bit surprised you're getting 4 chained mblks with a small MTU, but I don't really have any real evidence for that intuition. The fix is definitely to just _not unwrap_, which I'm starting on now.

**Comment by rcgoodfellow** [comment]
Dropping in the MDB output from our group debugging session.

[mdb-session.txt](https://github.com/oxidecomputer/opte/files/10772244/mdb-session.txt)

The core dump is rather large. Contact me if you want it.

**Comment by bnaecker** [comment]
Dropping some notes from a chat with Robert.

We're currently using a `heapless::Vec` for the `PacketSeg`s within a `Packet`. That's statically-sized to 4, meaning we can't handle packets split into more than 4 `mblk_t`s, linked by their `b_cont` pointers. That's exactly what we hit here. Also, there is no guaranteed limit on the number of blocks chained together, so _any_ statically-sized array is unlikely to work.

At the same time, we'd _really_ like to avoid any allocations here. That means we'll likely need to change the `heapless::Vec` into a true linked-list. I don't see any obvious way to avoid that, and it maps most directly to the underlying structure of the `mblk_t` that's being wrapped.

**Comment by bnaecker** [comment]
Closed by #336
